readme.txt
Able Avionics - Sandwich edition is for implementing the new hullmod on Garou Labs ships

changelog.txt
-Any version change that isn't listed only includes compatibility changes

1.3.1
- 0.96 compatibility

1.3.0
- update for diable 2.7.0 (hull changes n shit carried over)
- versant shields to 0.7 (fck u)
- fractus gains 15 OP instead of gantry

1.2.0
- variant updates for 0.95.1a

1.1.0
- changed prefix of hullmod to avoid confusion
- breaks saves, your save is gonna break if you update unless you replace instances of the hullmod with the new ID in your save

1.0.3
- panda regains part of its shield buffs post nerfs

1.0.2
- adjustments post additional panda changes

1.0.1
- updated for diable 2.63
- panda shield back to 1.0, built in expanded decks removed (due to massive buffs)
- no longer removes versant hullmod, since base diable now only blocks SO

1.0.0
- initial release

src/data/hullmods/AbleAvionicsTuning.java
package data.hullmods;

import com.fs.starfarer.api.Global;
import com.fs.starfarer.api.campaign.CargoStackAPI;
import com.fs.starfarer.api.combat.BaseHullMod;
import com.fs.starfarer.api.combat.MutableShipStatsAPI;
import com.fs.starfarer.api.combat.ShipAPI;
import com.fs.starfarer.api.combat.ShipAPI.HullSize;
import java.util.HashMap;
import java.util.Map;
import org.lazywizard.lazylib.MathUtils;
import static data.scripts.util.Diableavionics_stringsManager.txt;

public class AbleAvionicsTuning extends BaseHullMod {
    
/*    private static float debuff=0;*/
    
/*    private static final Map<String,Float> HULLMOD_DEBUFF = new HashMap<>();
    static{
        HULLMOD_DEBUFF.put("safetyoverrides",0.2f);
        HULLMOD_DEBUFF.put("unstable_injector",0.15f);
        HULLMOD_DEBUFF.put("auxiliarythrusters",0.15f);
        HULLMOD_DEBUFF.put("SCY_lightArmor",0.15f);
    }*/
    
    private final Map<Integer,String> LEFT_SELECTOR = new HashMap<>();
    {
        LEFT_SELECTOR.put(0, "diableavionics_versant_harvest_LEFT");
        LEFT_SELECTOR.put(1, "diableavionics_versant_harvestB_LEFT");
        LEFT_SELECTOR.put(2, "diableavionics_versant_harvestC_LEFT");
    }
    
    private final Map<Integer,String> RIGHT_SELECTOR = new HashMap<>();
    {
        RIGHT_SELECTOR.put(0, "diableavionics_versant_harvest_RIGHT");
        RIGHT_SELECTOR.put(1, "diableavionics_versant_harvestB_RIGHT");
        RIGHT_SELECTOR.put(2, "diableavionics_versant_harvestC_RIGHT");
    }
    
    private final Map<String, Integer> SWITCH_TO = new HashMap<>();
    {
        SWITCH_TO.put("diableavionics_versant_harvest_LEFT",1);
        SWITCH_TO.put("diableavionics_versant_harvestB_LEFT",2);
        SWITCH_TO.put("diableavionics_versant_harvestC_LEFT",0);
    }
    
    private final Map<Integer,String> SWITCH = new HashMap<>();
    {
        SWITCH.put(0,"diableavionics_selector_auto");
        SWITCH.put(1,"diableavionics_selector_burst");
        SWITCH.put(2,"diableavionics_selector_semi");
    }    
    
    private final String leftslotID = "GUN_LEFT"; 
    private final String rightslotID = "GUN_RIGHT";     
    
    @Override
    public void applyEffectsBeforeShipCreation(HullSize hullSize, MutableShipStatsAPI stats, String id) {
/*        debuff=0;
        for(String h : stats.getVariant().getHullMods()){
            if(HULLMOD_DEBUFF.containsKey(h)){
                debuff+=HULLMOD_DEBUFF.get(h);
            }
        }
        stats.getPeakCRDuration().modifyMult(id,1-debuff);*/
        
        //trigger a weapon switch if none of the selector hullmods are present
        boolean toSwitch=true;
        for(int i=0; i<SWITCH.size(); i++){
            if(stats.getVariant().getHullMods().contains(SWITCH.get(i))){
                toSwitch=false;
            }
        }
        
        //remove the weapons to change and swap the hullmod for the next fire mode
        if(toSwitch){        
            //select new fire mode
            int selected;       
            boolean random=false;
            if(stats.getVariant().getWeaponSpec(leftslotID)!=null){
                selected=SWITCH_TO.get(stats.getVariant().getWeaponSpec(leftslotID).getWeaponId());
                
            } else {
                selected=MathUtils.getRandomNumberInRange(0, SWITCH_TO.size()-1);
                random=true;
            }
            
            //add the proper hullmod
            stats.getVariant().addMod(SWITCH.get(selected));

            //clear the weapons to replace
            stats.getVariant().clearSlot(leftslotID);
            stats.getVariant().clearSlot(rightslotID);
            //select and place the proper weapon
            String toInstallLeft=LEFT_SELECTOR.get(selected);                
            String toInstallRight=RIGHT_SELECTOR.get(selected);

            stats.getVariant().addWeapon(leftslotID, toInstallLeft);
            stats.getVariant().addWeapon(rightslotID, toInstallRight);
            
            if(random){
                stats.getVariant().autoGenerateWeaponGroups();
            }
        }
    }
    
    @Override
    public void applyEffectsAfterShipCreation(ShipAPI ship, String id){
        //only check for undo in refit to avoid issues
        if(ship.getOriginalOwner()<0){
            //undo fix for harvests put in cargo
            if(
                    Global.getSector()!=null && 
                    Global.getSector().getPlayerFleet()!=null && 
                    Global.getSector().getPlayerFleet().getCargo()!=null && 
                    Global.getSector().getPlayerFleet().getCargo().getStacksCopy()!=null &&
                    !Global.getSector().getPlayerFleet().getCargo().getStacksCopy().isEmpty()
                    ){
                for (CargoStackAPI s : Global.getSector().getPlayerFleet().getCargo().getStacksCopy()){
                    if(
                            s.isWeaponStack() && (
                                LEFT_SELECTOR.containsValue(s.getWeaponSpecIfWeapon().getWeaponId()) || 
                                RIGHT_SELECTOR.containsValue(s.getWeaponSpecIfWeapon().getWeaponId())
                                ) 
                            ){
                        Global.getSector().getPlayerFleet().getCargo().removeStack(s);
                    }
                }
            }
        }
    }
    
    @Override
    public String getDescriptionParam(int index, HullSize hullSize) { 
        if (index == 0) return txt("hm_tuning_0");
        if (index == 1) return txt("hm_tuning_1");
        if (index == 2) return txt("hm_tuning_2");
        if (index == 3) return txt("hm_tuning_3");        
        
        //if (index == 4) return (int)(-debuff*100)+txt("%");
        return null;
    }
    
    @Override
    public boolean isApplicableToShip(ShipAPI ship) {
        // Allows any ship with a diableavionics hull id
        return ( ship.getHullSpec().getHullId().startsWith("diableavionics_"));	
    }
}


src/data/hullmods/DiableAvionicsUpgrade_Reborn.java
package data.hullmods;

import com.fs.starfarer.api.Global;
import com.fs.starfarer.api.combat.BaseHullMod;
import com.fs.starfarer.api.combat.ShipAPI;
import com.fs.starfarer.api.combat.ShipAPI.HullSize;
import com.fs.starfarer.api.combat.WeaponAPI.WeaponType;

import java.awt.*;
import java.util.EnumSet;


public class DiableAvionicsUpgrade_Reborn extends BaseHullMod {

    private final float CHECK=1f;
    private float timer=0, previous=0;
    private final String ID="Target Analysis";

    
        private static class BuffData {
        float increment;
        float autoaimMin;
        float autoaimMax;
        float accelBuff;
        float turnBuff;
        float weaponTurn;
        float shieldUnfold;
        float shieldTurn;

        public BuffData(float increment, float autoaimMin, float autoaimMax, float accelBuff, float turnBuff, float weaponTurn, float shieldUnfold, float shieldTurn) {
            this.increment = increment;
            //the effect ranges from 0 to 1. ie an increment of 1 goes straight from 0 to 1, while an increment of 0.25 needs 4 steps.
            this.autoaimMin = autoaimMin;
            this.autoaimMax = autoaimMax;
            this.accelBuff = accelBuff;
            this.turnBuff = turnBuff;
            this.weaponTurn = weaponTurn;
            this.shieldUnfold = shieldUnfold;
            this.shieldTurn = shieldTurn;
        }
    }    
    
    private static final BuffData powerBuff = new BuffData(1f/4f, -15, +100, 100, 100, 50,
            100, 100);

    
    @Override
    public void advanceInCombat(ShipAPI ship, float amount) {
        timer=Global.getCombatEngine().getTotalElapsedTime(false);
        
        ShipAPI playerShip = Global.getCombatEngine().getPlayerShip();
        if(previous>=timer || timer>previous+CHECK || ship==playerShip){
            
            if(ship!=playerShip){
                previous=timer;
            }

            int tick;

            switch (ship.getHullSize()) {
                case FRIGATE:
                    tick = (int)ship.getTimeDeployedForCRReduction()/30;
                    break;
                case DESTROYER:
                    tick = (int)ship.getTimeDeployedForCRReduction()/45;
                    break;
                case CRUISER:
                    tick = (int)ship.getTimeDeployedForCRReduction()/60;
                    break;
                case CAPITAL_SHIP:
                    tick = (int)ship.getTimeDeployedForCRReduction()/75;
                    break;
                default:
                    throw new IllegalStateException("Unexpected value: " + ship.getHullSize());
            }


            float effectLevel = Math.min(1, (powerBuff.increment)*tick);
            float autoaimEffect = (powerBuff.autoaimMin) + effectLevel * ((powerBuff.autoaimMax)-(powerBuff.autoaimMin));

            ship.getMutableStats().getWeaponTurnRateBonus().modifyMult(ID, 1+((powerBuff.weaponTurn)*effectLevel/100));            
            ship.getMutableStats().getAutofireAimAccuracy().modifyMult(ID, 1+(autoaimEffect/100));
            ship.getMutableStats().getRecoilDecayMult().modifyMult(ID, 1+(autoaimEffect/100));
            ship.getMutableStats().getRecoilPerShotMult().modifyMult(ID, 1-(autoaimEffect/100));   

            ship.getMutableStats().getAcceleration().modifyMult(ID, 1+((powerBuff.accelBuff)*effectLevel/100));
            ship.getMutableStats().getDeceleration().modifyMult(ID, 1+((powerBuff.accelBuff)*effectLevel/100));
            ship.getMutableStats().getTurnAcceleration().modifyMult(ID, 1+((powerBuff.turnBuff)*effectLevel/100));

            ship.getMutableStats().getShieldTurnRateMult().modifyMult(ID, 1+((powerBuff.shieldTurn)*effectLevel/100));
            ship.getMutableStats().getShieldUnfoldRateMult().modifyMult(ID, 1+((powerBuff.shieldUnfold)*effectLevel/100));

            if (effectLevel >= 1) {
                ship.getMutableStats().getBallisticWeaponDamageMult().modifyMult(ID, 1.1f);
                ship.getMutableStats().getEnergyWeaponDamageMult().modifyMult(ID, 1.1f);
            }

            EnumSet WEAPON_TYPES = EnumSet.of(WeaponType.BALLISTIC,WeaponType.ENERGY);

/*            if (0.01f < effectLevel && effectLevel <  0.25f){
                ship.setWeaponGlow(
                    1f,
                        new Color(0, 127, 0, 64),
                        WEAPON_TYPES);
            }else if (effectLevel < 0.5f){
                ship.setWeaponGlow(
                        0f,
                        new Color(0,0,0),
                        WEAPON_TYPES);
            }else*/ if (effectLevel == 0.75f){
                ship.setWeaponGlow(
                        0.75f,
                        new Color(255,127,0, 128),
                        WEAPON_TYPES);
            }else if (effectLevel == 1f){
                ship.setWeaponGlow(
                        1f,
                        new Color(255,0,0,128),
                        WEAPON_TYPES);
            }

            if(ship==playerShip){
                Global.getCombatEngine().maintainStatusForPlayerShip(
                        "AdvancedAvionicsBoost",
                        "graphics/icons/hullsys/high_energy_focus.png",
                        "Pattern Analysis:",
                        Math.round(effectLevel*100)+"% complete.",
                        effectLevel < 0.25f);
            }
        }
    }
    
    @Override
    public String getDescriptionParam(int index, HullSize hullSize) {
        if (index == 0) {
            return "-15%";
        }
        if (index == 1) {
            return "30/45/60/75 seconds";
        }
        if (index == 2) {
            return "120/180/240/300 seconds";
        }
        if (index == 3) {
            return "100%";
        }
        if (index == 4) {
            return "100%";
        }
        if (index == 5) {
            return "100%";
        }
        if (index == 6) {
            return "10%";
        }
        return null;
    }

}


src/data/hullmods/DiableAvionicsTuning_butCooler.java
package data.hullmods;

import com.fs.starfarer.api.Global;
import com.fs.starfarer.api.campaign.CargoStackAPI;
import com.fs.starfarer.api.combat.BaseHullMod;
import com.fs.starfarer.api.combat.MutableShipStatsAPI;
import com.fs.starfarer.api.combat.ShipAPI;
import com.fs.starfarer.api.combat.ShipAPI.HullSize;
import java.util.HashMap;
import java.util.Map;
import org.lazywizard.lazylib.MathUtils;
import static data.scripts.util.Diableavionics_stringsManager.txt;

public class DiableAvionicsTuning_butCooler extends BaseHullMod {
    
/*    private static float debuff=0;*/
    
/*    private static final Map<String,Float> HULLMOD_DEBUFF = new HashMap<>();
    static{
        HULLMOD_DEBUFF.put("safetyoverrides",0.2f);
        HULLMOD_DEBUFF.put("unstable_injector",0.15f);
        HULLMOD_DEBUFF.put("auxiliarythrusters",0.15f);
        HULLMOD_DEBUFF.put("SCY_lightArmor",0.15f);
    }*/
    
    private final Map<Integer,String> LEFT_SELECTOR = new HashMap<>();
    {
        LEFT_SELECTOR.put(0, "diableavionics_versant_harvest_LEFT");
        LEFT_SELECTOR.put(1, "diableavionics_versant_harvestB_LEFT");
        LEFT_SELECTOR.put(2, "diableavionics_versant_harvestC_LEFT");
    }
    
    private final Map<Integer,String> RIGHT_SELECTOR = new HashMap<>();
    {
        RIGHT_SELECTOR.put(0, "diableavionics_versant_harvest_RIGHT");
        RIGHT_SELECTOR.put(1, "diableavionics_versant_harvestB_RIGHT");
        RIGHT_SELECTOR.put(2, "diableavionics_versant_harvestC_RIGHT");
    }
    
    private final Map<String, Integer> SWITCH_TO = new HashMap<>();
    {
        SWITCH_TO.put("diableavionics_versant_harvest_LEFT",1);
        SWITCH_TO.put("diableavionics_versant_harvestB_LEFT",2);
        SWITCH_TO.put("diableavionics_versant_harvestC_LEFT",0);
    }
    
    private final Map<Integer,String> SWITCH = new HashMap<>();
    {
        SWITCH.put(0,"diableavionics_selector_auto");
        SWITCH.put(1,"diableavionics_selector_burst");
        SWITCH.put(2,"diableavionics_selector_semi");
    }    
    
    private final String leftslotID = "GUN_LEFT"; 
    private final String rightslotID = "GUN_RIGHT";     
    
    @Override
    public void applyEffectsBeforeShipCreation(HullSize hullSize, MutableShipStatsAPI stats, String id) {
/*        debuff=0;
        for(String h : stats.getVariant().getHullMods()){
            if(HULLMOD_DEBUFF.containsKey(h)){
                debuff+=HULLMOD_DEBUFF.get(h);
            }
        }
        stats.getPeakCRDuration().modifyMult(id,1-debuff);*/
        
        //trigger a weapon switch if none of the selector hullmods are present
        boolean toSwitch=true;
        for(int i=0; i<SWITCH.size(); i++){
            if(stats.getVariant().getHullMods().contains(SWITCH.get(i))){
                toSwitch=false;
            }
        }
        
        //remove the weapons to change and swap the hullmod for the next fire mode
        if(toSwitch){        
            //select new fire mode
            int selected;       
            boolean random=false;
            if(stats.getVariant().getWeaponSpec(leftslotID)!=null){
                selected=SWITCH_TO.get(stats.getVariant().getWeaponSpec(leftslotID).getWeaponId());
                
            } else {
                selected=MathUtils.getRandomNumberInRange(0, SWITCH_TO.size()-1);
                random=true;
            }
            
            //add the proper hullmod
            stats.getVariant().addMod(SWITCH.get(selected));

            //clear the weapons to replace
            stats.getVariant().clearSlot(leftslotID);
            stats.getVariant().clearSlot(rightslotID);
            //select and place the proper weapon
            String toInstallLeft=LEFT_SELECTOR.get(selected);                
            String toInstallRight=RIGHT_SELECTOR.get(selected);

            stats.getVariant().addWeapon(leftslotID, toInstallLeft);
            stats.getVariant().addWeapon(rightslotID, toInstallRight);
            
            if(random){
                stats.getVariant().autoGenerateWeaponGroups();
            }
        }
    }
    
    @Override
    public void applyEffectsAfterShipCreation(ShipAPI ship, String id){
        //only check for undo in refit to avoid issues
        if(ship.getOriginalOwner()<0){
            //undo fix for harvests put in cargo
            if(
                    Global.getSector()!=null && 
                    Global.getSector().getPlayerFleet()!=null && 
                    Global.getSector().getPlayerFleet().getCargo()!=null && 
                    Global.getSector().getPlayerFleet().getCargo().getStacksCopy()!=null &&
                    !Global.getSector().getPlayerFleet().getCargo().getStacksCopy().isEmpty()
                    ){
                for (CargoStackAPI s : Global.getSector().getPlayerFleet().getCargo().getStacksCopy()){
                    if(
                            s.isWeaponStack() && (
                                LEFT_SELECTOR.containsValue(s.getWeaponSpecIfWeapon().getWeaponId()) || 
                                RIGHT_SELECTOR.containsValue(s.getWeaponSpecIfWeapon().getWeaponId())
                                ) 
                            ){
                        Global.getSector().getPlayerFleet().getCargo().removeStack(s);
                    }
                }
            }
        }
    }
    
    @Override
    public String getDescriptionParam(int index, HullSize hullSize) { 
        if (index == 0) return txt("hm_tuning_0");
        if (index == 1) return txt("hm_tuning_1");
        if (index == 2) return txt("hm_tuning_2");
        if (index == 3) return txt("hm_tuning_3");        
        
        //if (index == 4) return (int)(-debuff*100)+txt("%");
        return null;
    }
    
    @Override
    public boolean isApplicableToShip(ShipAPI ship) {
        // Allows any ship with a diableavionics hull id
        return ( ship.getHullSpec().getHullId().startsWith("diableavionics_"));	
    }
}


src/data/hullmods/AbleAvionicsUpgrade_Reborn.java
package data.hullmods;

import com.fs.starfarer.api.Global;
import com.fs.starfarer.api.combat.BaseHullMod;
import com.fs.starfarer.api.combat.ShipAPI;
import com.fs.starfarer.api.combat.ShipAPI.HullSize;
import com.fs.starfarer.api.combat.WeaponAPI.WeaponType;

import java.awt.*;
import java.util.EnumSet;


public class AbleAvionicsUpgrade_Reborn extends BaseHullMod {

    private final float CHECK=1f;
    private float timer=0, previous=0;
    private final String ID="Target Analysis";

    
        private static class BuffData {
        float increment;
        float autoaimMin;
        float autoaimMax;
        float accelBuff;
        float turnBuff;
        float weaponTurn;
        float shieldUnfold;
        float shieldTurn;

        public BuffData(float increment, float autoaimMin, float autoaimMax, float accelBuff, float turnBuff, float weaponTurn, float shieldUnfold, float shieldTurn) {
            this.increment = increment;
            //the effect ranges from 0 to 1. ie an increment of 1 goes straight from 0 to 1, while an increment of 0.25 needs 4 steps.
            this.autoaimMin = autoaimMin;
            this.autoaimMax = autoaimMax;
            this.accelBuff = accelBuff;
            this.turnBuff = turnBuff;
            this.weaponTurn = weaponTurn;
            this.shieldUnfold = shieldUnfold;
            this.shieldTurn = shieldTurn;
        }
    }    
    
    private static final BuffData powerBuff = new BuffData(1f/4f, -15, +100, 100, 100, 50,
            100, 100);

    
    @Override
    public void advanceInCombat(ShipAPI ship, float amount) {
        timer=Global.getCombatEngine().getTotalElapsedTime(false);
        
        ShipAPI playerShip = Global.getCombatEngine().getPlayerShip();
        if(previous>=timer || timer>previous+CHECK || ship==playerShip){
            
            if(ship!=playerShip){
                previous=timer;
            }

            int tick;

            switch (ship.getHullSize()) {
                case FRIGATE:
                    tick = (int)ship.getTimeDeployedForCRReduction()/30;
                    break;
                case DESTROYER:
                    tick = (int)ship.getTimeDeployedForCRReduction()/45;
                    break;
                case CRUISER:
                    tick = (int)ship.getTimeDeployedForCRReduction()/60;
                    break;
                case CAPITAL_SHIP:
                    tick = (int)ship.getTimeDeployedForCRReduction()/75;
                    break;
                default:
                    throw new IllegalStateException("Unexpected value: " + ship.getHullSize());
            }


            float effectLevel = Math.min(1, (powerBuff.increment)*tick);
            float autoaimEffect = (powerBuff.autoaimMin) + effectLevel * ((powerBuff.autoaimMax)-(powerBuff.autoaimMin));

            ship.getMutableStats().getWeaponTurnRateBonus().modifyMult(ID, 1+((powerBuff.weaponTurn)*effectLevel/100));            
            ship.getMutableStats().getAutofireAimAccuracy().modifyMult(ID, 1+(autoaimEffect/100));
            ship.getMutableStats().getRecoilDecayMult().modifyMult(ID, 1+(autoaimEffect/100));
            ship.getMutableStats().getRecoilPerShotMult().modifyMult(ID, 1-(autoaimEffect/100));   

            ship.getMutableStats().getAcceleration().modifyMult(ID, 1+((powerBuff.accelBuff)*effectLevel/100));
            ship.getMutableStats().getDeceleration().modifyMult(ID, 1+((powerBuff.accelBuff)*effectLevel/100));
            ship.getMutableStats().getTurnAcceleration().modifyMult(ID, 1+((powerBuff.turnBuff)*effectLevel/100));

            ship.getMutableStats().getShieldTurnRateMult().modifyMult(ID, 1+((powerBuff.shieldTurn)*effectLevel/100));
            ship.getMutableStats().getShieldUnfoldRateMult().modifyMult(ID, 1+((powerBuff.shieldUnfold)*effectLevel/100));

            if (effectLevel >= 1) {
                ship.getMutableStats().getBallisticWeaponDamageMult().modifyMult(ID, 1.1f);
                ship.getMutableStats().getEnergyWeaponDamageMult().modifyMult(ID, 1.1f);
            }

            EnumSet<WeaponType> WEAPON_TYPES = EnumSet.of(WeaponType.BALLISTIC,WeaponType.ENERGY);

/*            if (0.01f < effectLevel && effectLevel <  0.25f){
                ship.setWeaponGlow(
                    1f,
                        new Color(0, 127, 0, 64),
                        WEAPON_TYPES);
            }else if (effectLevel < 0.5f){
                ship.setWeaponGlow(
                        0f,
                        new Color(0,0,0),
                        WEAPON_TYPES);
            }else*/ if (effectLevel == 0.75f){
                ship.setWeaponGlow(
                        0.75f,
                        new Color(255,127,0, 128),
                        WEAPON_TYPES);
            }else if (effectLevel == 1f){
                ship.setWeaponGlow(
                        1f,
                        new Color(255,0,0,128),
                        WEAPON_TYPES);
            }

            if(ship==playerShip){
                Global.getCombatEngine().maintainStatusForPlayerShip(
                        "AdvancedAvionicsBoost",
                        "graphics/icons/hullsys/high_energy_focus.png",
                        "Pattern Analysis:",
                        Math.round(effectLevel*100)+"% complete.",
                        effectLevel < 0.25f);
            }
        }
    }
    
    @Override
    public String getDescriptionParam(int index, HullSize hullSize) {
        if (index == 0) {
            return "-15%";
        }
        if (index == 1) {
            return "30/45/60/75 seconds";
        }
        if (index == 2) {
            return "120/180/240/300 seconds";
        }
        if (index == 3) {
            return "100%";
        }
        if (index == 4) {
            return "100%";
        }
        if (index == 5) {
            return "100%";
        }
        if (index == 6) {
            return "10%";
        }
        return null;
    }

}
